src/ServiceBus.AttachmentPlugin/AzureStorageAttachment.cs

```diff
@@ -145,5 +151,29 @@ static void SetValidUntil(ICloudBlob blob, TimeSpan timeToBeReceived)
             message.Body = bytes;
             return message;
         }
+
+        void ThrowIfDisposed()
+        {
+            if (disposed)
+            {
+                throw new ObjectDisposedException($"{nameof(AzureStorageAttachment)} has been already disposed.");
+            }
+        }
+
+        public void Dispose()
+        {
+            if (Interlocked.Exchange(ref disposeSignaled, 1) != 0)
+            {
+                return;
+            }
+
+            var semaphorToDispose = Interlocked.Exchange(ref semaphore, null);
```

Comment:
my understanding was it was safe to use null propagation for dispose of fields. ie instead of `Interlocked.Exchange` just do `semaphore?.Dispose()`

Reply:
Nice. Makes it much simpler 🙂